### ISSUE TITLE
Inference performance gap between flash attention and Pytorch

### ISSUE THREAD

**Issue by Tengfei09** [HazyResearch/flash-attention#36]
I have a question about the inference performance gap between this project and the standard attention used in PyTorch.

After reading your python front-end API, I found that flash attention will remove the padding of sentences to prevent computing the useless tokens(unpadding), while PyTorch does not. is it a fair comparison? 

I mean, what is the root cause of the performance improvement? unpadding or kernel fusion(batchGEMM + Softmax + batchGEMM) used in flash attention?

**Comment by tridao** [comment]
The root cause of performance improvement is reduced reads/writes to GPU HBM (kernel fusion is a way to get there). For example, in the GPT2 benchmark, there's no padding/unpadding.
Unpadding is an extra optimization, and Pytorch has also started implementing it (called [NestedTensor](https://pytorch.org/tutorials/prototype/nestedtensor.html)).

**Comment by Tengfei09** [comment]
By the way, could you provide a PyTorch-independent Cuda kernel implementation? I'd like to export this code into a TensorFlow version.
